### PR TITLE
Cbrune/bfd 525 paging bug

### PR DESCRIPTION
**Pull request by cbrunefearless**
### Change Details
I believe there are three issues with using the header "IncludeIdentifiers=mbi" on the Patient endpoint to search for contract membership by month.

Next link (for paging through lists of results) is not being included on many (maybe all) production contracts as part of the response to a search.
An extra empty next link is added in specific conditions. When queried it throws a 500 error.
If there is no membership information for a given month a 500 error is thrown like with issue 2.
The exact same calls without using "IncludeIdentifiers=mbi" as a header succeed.

Our PDP partners are requesting this information and need it to proceed.

This change fixes issues from the next link (for paging through lists of results) that is not being included on many production contracts as part of the response to a search.  Also adds the next link in specific conditions.

### Acceptance Validation

Unit tests are included with fix and IT tests pass.

### Feedback Requested

The changes make sense, unit tests pass, and IT tests pass as well.
### External References

- [BLUEBUTTON-xxx](https://jira.cms.gov/browse/BFD-525)

### Security Implications

- [ ] new software dependencies

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

**Review by wnyffenegger — DISMISSED** [review]

**Review by nflearl — DISMISSED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 1]
```diff
@@ -269,6 +269,13 @@ public Bundle searchByCoverageContract(
     List<Beneficiary> matchingBeneficiaries =
         fetchBeneficiaries(coverageId, includeIdentifiersValues, paging);
 
+    // Insures backwards compatability
+    boolean hasAnotherPage = matchingBeneficiaries.size() == paging.getPageSize();
```
Comment (karlmdavis) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:273]: Definitely looks broken, and doesn't match the V1 logic.
```suggestion
    boolean hasAnotherPage = matchingBeneficiaries.size() > paging.getPageSize();
```

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 2]
```diff
@@ -454,7 +459,7 @@ private String partDFieldFor(CcwCodebookVariable month) {
     if (hasHICN(identifiers)) joinsClause += "left join fetch b.beneficiaryHistories ";
 
     String query =
-        "select b from Beneficiary b "
+        "select distinct b from Beneficiary b "
```
Comment (karlmdavis) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:462]: Just a note: I haven't tested this at all, so please make sure you do.

Comment (karlmdavis) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:462]: This same change would also be needed in the V2 resource provider.

**Review by karlmdavis — CHANGES_REQUESTED** [review]
Solid fixes! 🏅 

Found a couple of bugs that need to be fixed, and it may still need some testing.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java (hunk 1 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java]: fix that

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java (hunk 2 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:462]: fix that

Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:462]: fix that

**Review by nflearl — DISMISSED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java (hunk 2 above)
Comment (karlmdavis) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:462]: ?

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientLinkBuilderTest.java [hunk 3]
```diff
@@ -103,6 +103,28 @@ public void fullPageTest() {
     Assert.assertTrue(bundle.getLink().isEmpty());
     paging.addLinks(bundle);
 
+    Assert.assertNotNull(bundle.getLink(Constants.LINK_SELF));
+    Assert.assertNotNull(bundle.getLink(Constants.LINK_FIRST));
+    Assert.assertNull(bundle.getLink(Constants.LINK_NEXT));
+  }
+
+  @Test
+  public void fullPageTestWithPaging() {
```
Comment (karlmdavis) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientLinkBuilderTest.java:112]: I think you should consider adding a unit test for the `distinct` issue, if you can. Not a hard requirement; just a suggestion.

**Review by karlmdavis — APPROVED** [review]
LGTM, though I do have one suggestion to consider.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 4]
```diff
@@ -269,6 +269,13 @@ public Bundle searchByCoverageContract(
     List<Beneficiary> matchingBeneficiaries =
         fetchBeneficiaries(coverageId, includeIdentifiersValues, paging);
 
+    // Insures backwards compatability
```
Comment (nflearl) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:272]: Delete this comment as when we originally modified this file, we did not apply the +1 fix.  Now that the fix has been applied, this comment no longer applies.

**Review by nflearl — COMMENTED** [review]
The backwards compatible comment no longer applies.

**Review by wnyffenegger — APPROVED** [review]

**Review by nflearl — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientLinkBuilderTest.java (hunk 3 above)
Comment (karlmdavis) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientLinkBuilderTest.java:112]: FYI: We took care of this on Zoom last Friday.

**Review by karlmdavis — COMMENTED** [review]
LGTM 🎸

**Review by karlmdavis — APPROVED** [review]
LGTM 🎸  x 2

(Must have had the wrong option selected, earlier. Apologies!)

**Pull request merged**